sklearn/ensemble/tests/test_gradient_boosting.py

```diff
@@ -1106,3 +1115,41 @@ def test_sparse_input():
 
     for EstimatorClass, sparse_matrix in product(ests, sparse_matrices):
         yield check_sparse_input, EstimatorClass, X, sparse_matrix(X), y
+
+
+def test_gradient_boosting_early_stopping():
+    X_clf, y_clf = make_classification(n_samples=1000, random_state=0)
+    X_reg, y_reg = X_clf, y_clf
+
+    # Check if early_stopping works as expected
+    for (tol, final_n_est_clf, final_n_est_reg) in ((1e-1, 24, 13),
```

Comment:
I assume these aren't hand-calculated. Perhaps add a comment to say how these numbers were derived.

I'd also prefer something like:

```python
gbc = GradientBoostingClassifier(n_estimators=100, learning_rate=0.1, max_depth=3, random_state=42)
gbr = ...

assert gbc.fit(X_train, y_train).n_estimators_ == 100
assert gbr.fit(X_train, y_train).n_estimators_ == 100

for est, tol, early_stop_n_estimators in [(gbc, 1e-1, 24), (gbr, 1e-1, 13), ...]:
    est.set_params(tol=tol, n_iter_no_change=10)
    est.fit(X_train, y_train)
    ...
```